src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingRequestResolver.java

```diff
@@ -0,0 +1,140 @@
+        return attemptResolveForBuildingAndStack(manager, buildingWorker, request.getRequest().getStack());
+    }
+
+    @Nullable
+    protected List<IToken<?>> attemptResolveForBuildingAndStack(@NotNull final IRequestManager manager, @NotNull final AbstractBuildingWorker building, final ItemStack stack)
+    {
+        final IRecipeStorage fullfillableCrafting = building.getFirstFullFillableRecipe(stack);
+        if (fullfillableCrafting != null)
+        {
+            return ImmutableList.of();
+        }
+
+        final IRecipeStorage craftableCrafting = building.getFirstRecipe(stack);
+        if (craftableCrafting == null)
+        {
+            return null;
+        }
+
+        return createRequestsForRecipe(manager, building, stack, craftableCrafting);
+    }
+
+    protected int calculateMaxCraftingCount(@NotNull final ItemStack outputStack, @NotNull final IRecipeStorage storage)
+    {
+        //Calculate the initial crafting count from the request and the storage output.
+        int craftingCount =
+          (int) Math.ceil(Math.max(ItemStackUtils.getSize(outputStack), ItemStackUtils.getSize(storage.getPrimaryOutput())) / ItemStackUtils.getSize(storage.getPrimaryOutput()));
+
+        //Now check if we excede an ingredients max stack size.
+        for(final ItemStack ingredient : storage.getInput())
+        {
+            //Calculate the input count for the ingredient.
+            final int ingredientInputCount = ItemStackUtils.getSize(ingredient) * craftingCount;
+            //Check if we are above the max stacksize.
+            if (ingredientInputCount > ingredient.getMaxStackSize())
+            {
+                //Recalculate the crafting limit using the maxstacksize of the ingredient.
+                craftingCount =
+                  (int) Math.ceil(Math.max(ingredient.getMaxStackSize(), ItemStackUtils.getSize(storage.getPrimaryOutput())) / ItemStackUtils.getSize(storage.getPrimaryOutput()));
+            }
+        }
+
+        return craftingCount;
+    }
+
+    @Nullable
+    protected List<IToken<?>> createRequestsForRecipe(@NotNull final IRequestManager manager, @NotNull final AbstractBuildingWorker building, final ItemStack requestStack, @NotNull final IRecipeStorage storage)
+    {
+        final int craftingCount = calculateMaxCraftingCount(requestStack, storage);
+        return storage.getInput().stream()
+                 .filter(s -> !ItemStackUtils.isEmpty(s))
+                 .map(stack -> {
+                    final ItemStack craftingHelperStack = stack.copy();
+                    ItemStackUtils.setSize(craftingHelperStack, ItemStackUtils.getSize(craftingHelperStack) * craftingCount);
+
+                    return createNewRequestForStack(manager, craftingHelperStack);
+                }).collect(Collectors.toList());
+    }
+
+    @Nullable
+    protected IToken<?> createNewRequestForStack(@NotNull final IRequestManager manager, final ItemStack stack)
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Remove usage of generic wildcard type. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1452)